### DISCUSSION TITLE
Experimental Edge Runtime - using dynamic routes + getServerSideProps isn't using the custom 404 page?

### DISCUSSION THREAD

**Discussion by MichailiK** [vercel/next.js#51073] [Help]
### Summary

I'm using Next.js with the pages router. As I'm deploying this to Cloudflare Pages, I'm required to use the `experimental-edge` runtime for a dynamic route (in my minimal repro, it's `/blog/[id]`). When I use `getServerSideProps` and return `{ notFound: true }`, Next.js seems to end up using the _error page instead of the custom 404 page. When I set the runtime back to node, the 404 page is shown as expected.

In my minimal repro, this can be observed by:
- ✅ Attempting to access any non-existant page (i.e. `localhost:3000/non-existant`) - will show custom 404 page.
   
  ![image](https://github.com/vercel/next.js/assets/39029839/580315f1-d6f3-48b0-8cd6-fa9ed3ef67c2)

- ❌ Attempting to access a dynamic route that 404's (i.e. a blog post that doesn't exist, `localhost:3000/blog/asdfg`) - will show Next.js' _error page.
 
  ![image](https://github.com/vercel/next.js/assets/39029839/8fa99484-cc89-4551-991f-0988f01b4534)

I haven't been able to find any discussions about this, nor if this is a limitation of the Edge runtime or if this is a bug. I've decided to create a discussion instead of opening an issue.

Of course, this can be worked around in the meantime by making your own _error page. Although I'd prefer being able to use the 404 page, if possible.

### Additional information

```
$ next info
    Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 18.16.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.4.5-canary.9
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3

```

### Example

https://github.com/MichailiK/dynamic-routes-404-experimental-edge

**Comment by Curetix** [comment]
Same issue here. Took me a bit to figure out why it wasn't using my 404 page. For now I'm using the _error component like this:

```tsx
import { GetServerSidePropsContext, GetServerSidePropsResult } from "next";

import Error404 from "~/pages/404";
import Error500 from "~/pages/500";

type ErrorPageProps = {
  statusCode: number;
};

export async function getServerSideProps(
  context: GetServerSidePropsContext,
): Promise<GetServerSidePropsResult<ErrorPageProps>> {
  return {
    props: {
      statusCode: context.res.statusCode || 404,
    },
  };
}

export default function Error({ statusCode }: ErrorPageProps) {
  if (statusCode === 404) {
    return <Error404 />;
  } else {
    return <Error500 />;
  }
}

```

**Comment by klapec** [comment]
I've experienced the same behaviour on `next@14.0.1`.
